spec/fastsync/sandbox/fsPCBoundedBlocking.tla

```diff
@@ -0,0 +1,236 @@
+------------------------ MODULE fsPCBoundedBlocking ------------------------
+
+EXTENDS TLC, Integers, Sequences
+CONSTANTS MaxQueueSize
+None == -1  
+
+(*--algorithm message_queue
+variables
+  \* queues prefixed by A, B..so state is easy to inspect
+  Asend_q = <<>>,
+  Breceive_q = <<>>,
+  Cdemux_q = <<>>,
+  msg = [type |-> "none"],
+  peers = {"p1"},
+  done = FALSE, \* set to TRUE when demux is done, send and receive read this and stop
+  
+define
+  FilledQueue(queue) == Len(queue) >= MaxQueueSize
+  FilledQueues == 
+    /\ FilledQueue(Breceive_q)
+    /\ FilledQueue(Asend_q)
+    /\ FilledQueue(Cdemux_q) 
+end define;
+
+macro add_to_blocking_queue(queue, val) begin
+  await Len(queue) < MaxQueueSize;
+  queue := Append(queue, val);
+end macro;
+
+macro read_from_queue(queue) begin
+    await Len(queue) > 0;
+    \* write message in global variable, not possible to return values here or from procedure
+    msg := Head(queue);
+    queue := Tail(queue);
+end macro;
+    
+(* 
+<> bounded blocking queues
+              
+   -------------------------------------
+   |               peer                |
+   |  ..............................   |
+   |  |                            |   |
+   |  v                            |   |         
+   |send                        receive|
+   |  |                            ^   |
+   |  |                            |   |
+   ---+----------------------------+----
+      |                            |
+      |                            |
+   ---+----------------------------+----
+   |  |          this node         |   |
+   |  |                            |   |
+   |  v <> Breceive_q              ^   |
+   |receive                       send |
+   |  |                            ^ <> Asend_q
+   |  |                            |   |
+   |  v <> Cdemux_q (1000)         |   |
```

Comment:
So would this be the `events` channel in the reactor?  Or how do these three queues map to the Go?

Reply:
There is no mapping at this point...but yes, this would be the Events channel in Go. And the Asend_q the per reactor channel of Channel in MConnection...and this is not blocking in Go code (returns error on full) vs blocking here. 
Eventually we want to model different types of queues to map to blocking/ unblocking, bounded/ unbounded etc. Once we have these modules then we can play with different concurrency models. At this stage trying to figure out if PlusCal is good to code processes/ threads.